<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [63]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# достаём id из таблицы vacancies
query_3_1 = f'''select id 
            from vacancies  
            '''

In [5]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1.nunique() #считаем количество id

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\3125997171.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


id    49197
dtype: int64

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# достаём id из таблицы employers
query_3_2 = f'''select id
            from employers
            '''

In [7]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2.nunique() #считаем количество id

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\3114110936.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


id    23501
dtype: int64

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# достаём id из таблицы areas
query_3_3 = f'''select id
            from areas
            '''

In [9]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3.nunique() #считаем количество id 

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\2521518149.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


id    1362
dtype: int64

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# достаём id из таблицы industries
query_3_4 = f'''select id
            from industries
            '''

In [11]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4.nunique() #считаем количество id

#294

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\4289425953.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


id    294
dtype: int64

***

Вывод: 
- В базе данных содержится информация о 49197 вакансиях, что более чем в два раза превышает количество работодателей (23501). Значит, многие работодатели предлагают больше одной вакансии. 
- Вакансии представлены из 1362 регионов и относятся к 294 сферам.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# соединяем vacancies и areas, считаем id вакансий и группируем по названиям регионов
query_4_1 = f'''select count(v.id),
            a.name
            from vacancies v
            join areas as a on v.area_id = a.id
            group by a.name
            order by count(v.id) desc
            limit 5
            '''

In [13]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

#Москва, Санкт-Петербург, Минск, Новосибирск, Алматы

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\674693917.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


count             name
0   5333           Москва
1   2851  Санкт-Петербург
2   2112            Минск
3   2006      Новосибирск
4   1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# считаем количество строк в таблице vacancies, где заполнены поля с ЗП
query_4_2 = f'''select count(*)
            from vacancies v
            where salary_from is not null or salary_to is not null
            '''

In [15]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

#24073

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\1198068634.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# достаем из таблицы vacancies сведения о зарплате исчитаем среднее значение
query_4_3 = f'''select avg(salary_from),
            avg(salary_to)
            from vacancies v
            '''

In [17]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\3595259731.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


avg            avg
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# считаем количство строк в таблице vacancies, сгруппировав по schedule и employment
query_4_4 = f'''select count(*),
        schedule,
        employment
            from vacancies
            group by schedule, employment
            order by count(*) desc
            '''

In [19]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\2028972612.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# считаем id из vacancies, сгруппировав по experience
query_4_5 = f'''select count(id),
            experience as "Требуемый опыт работы"
            from vacancies
            group by experience
            order by count(id) asc
            '''

In [21]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\210463167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


count Требуемый опыт работы
0   1337           Более 6 лет
1   7197             Нет опыта
2  14511         От 3 до 6 лет
3  26152    От 1 года до 3 лет

***

Вывод:
- Самое большое количество вакансий ожидаемо относится к крупнейшим городам: Москва, СПб, Минск, Новосибирск, Алматы. 
- Работодатели ищут, главным образом, соискателей с опытом работы от 1 года до 3 лет и предлагают работу с полной занятостью. 
- При этом лишь около половины размещенных вакансий (24073) содержат информацию о заработной плате, которая находится в пределах от 71065	до 110537 рублей.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# объединяем таблицы vacancies и employer по id, считаем количество id, сгруппировав по названиям компаний 
query_5_1 = f'''select count(v.id),
            e.name
            from vacancies v
            join employers e on v.employer_id = e.id
            group by e.name
            order by count(v.id) desc
            limit 5
            '''

In [23]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\1147638897.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


count           name
0   1933         Яндекс
1    491     Ростелеком
2    444       Тинькофф
3    428           СБЕР
4    331  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# объединяем таблицы areas, employers и vacancies, считаем уникальные id работодателей и вакансий, сгруппировав по регионам
query_5_2 = f'''select a.name,
            count(distinct e.id) employers,
            count(distinct v.id) vacancies
            from areas a
            left join employers e on a.id = e.area
            left join vacancies v on v.area_id = a.id
            where v.id is null
            group by a.id
            order by employers desc
            '''

In [25]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\2306256317.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


name  employers  vacancies
0                Россия        410          0
1             Казахстан        207          0
2    Московская область         75          0
3    Краснодарский край         19          0
4    Ростовская область         18          0
..                  ...        ...        ...
588             Скидель          0          0
589          Белоозерск          0          0
590           Орловский          0          0
591          Павловская          0          0
592             Таксимо          0          0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# соединяем таблицы vacancies и employers, считаем уникальные id регионов и сортируем по названиям компаний
query_5_3 = f'''select e.name,
            count(distinct v.area_id) areas
            from vacancies v
            join employers e on v.employer_id = e.id
            group by e.name
            order by areas desc
            '''

In [27]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\133400711.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


name  areas
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# объединяем employers и employers_industries, и считаем те, где нет industry_id
query_5_4 = f'''select count(e.id)
            from employers e
            left join employers_industries ei on e.id = ei.employer_id
            where industry_id is null
            '''

In [29]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\1045119256.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# объединяем employers и employers_industries, и считаем те id, где industry_id = 4, сгруппировав по названию компании
query_5_5 = f'''select count(ei.industry_id),
            e.name
            from employers e
            left join employers_industries ei on e.id = ei.employer_id
            group by e.name
            having count(ei.industry_id) = 4
            order by e.name
            '''

In [31]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\4025176131.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


count             name
0         4     101 Интернет
1         4         21vek.by
2         4             2ГИС
3         4               2К
4         4      4 пикселя +
...     ...              ...
1131      4            ЮРИОН
1132      4  ЮТИП Технологии
1133      4           ЯКласс
1134      4            ЯрНео
1135      4    Яшин&Партнёры

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# employers, employers_industries, и industries по id, считаем id компаний, сгруппировав по названиям
query_5_6 = f'''select count(e.id)
            from employers e
            join employers_industries ei on e.id = ei.employer_id
            join industries i on ei.industry_id = i.id
            group by i.name
            having i.name like 'Разработка программного обеспечения'
            '''

In [35]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\2713946032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [165]:
!pip install lxml

                                              0.0/3.8 MB ? eta -:--:--
                                              0.0/3.8 MB 1.4 MB/s eta 0:00:03
                                              0.0/3.8 MB 1.4 MB/s eta 0:00:03
                                              0.1/3.8 MB 655.4 kB/s eta 0:00:06
     -                                        0.1/3.8 MB 547.6 kB/s eta 0:00:07
     -                                        0.1/3.8 MB 568.9 kB/s eta 0:00:07
     -                                        0.2/3.8 MB 573.4 kB/s eta 0:00:07
     --                                       0.2/3.8 MB 655.6 kB/s eta 0:00:06
     --                                       0.2/3.8 MB 577.6 kB/s eta 0:00:07
     --                                       0.3/3.8 MB 609.2 kB/s eta 0:00:06
     ---                                      0.3/3.8 MB 610.3 kB/s eta 0:00:06
     ---                                      0.4/3.8 MB 655.7 kB/s eta 0:00:06
     ----                                     0.4/3.8


[notice] A new release of pip is available: 23.1.2 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [60]:
# код для получения списка городов-милионников
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = tuple(pd.read_html(url)[0]['Город'])
cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [61]:
# vacancies, employers, areas, считаем количество вакансий в разных регионов по фильтру 'Яндекс' и полученному выше кортежу городов cities;
# далее считаем сумму вакансий и объединяем данные в одну таблицу
query_5_7 = f''' with temp as
            (select a.name as names,
            count(v.id) as count
            from vacancies v
            join employers e on v.employer_id = e.id
            join areas a on v.area_id = a.id
            where e.name like 'Яндекс' and a.name in {cities}
            group by a.name)
            select temp.names, temp.count
            from temp
            union all
            select 'Total', sum(temp.count)
            from temp
            
            '''

In [64]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\2595257991.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


names  count
0         Волгоград   24.0
1           Воронеж   32.0
2      Екатеринбург   39.0
3            Казань   25.0
4         Краснодар   30.0
5        Красноярск   23.0
6            Москва   54.0
7   Нижний Новгород   36.0
8       Новосибирск   35.0
9              Омск   21.0
10            Пермь   25.0
11   Ростов-на-Дону   25.0
12           Самара   26.0
13  Санкт-Петербург   42.0
14              Уфа   26.0
15        Челябинск   22.0
16            Total  485.0

***

Вывод:

- Основное количество размещенных вакансий принадлежит банкам и компаниям из отрасли информационных технологий. Больше всего вакансий относится компании "Яндекс", которая размещает вакансии в наибольшем количестве регионов (1933 вакансии в 181 регионе, из них 485 в регионах-миллионниках). 
- 8419 работодателей не указывают сферу деятельности.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [39]:
# считаем названия вакансий с 'данн' и 'data'
query_6_1 = f'''select count(v.name)
            from vacancies v
            where v.name ilike '%data%' 
            or v.name ilike '%данн%'
            '''

In [40]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\1281472032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [41]:
# считаем количесво вакансий с упомянутыми выше сочетаниями в названии и обязательно словом 'junior'
query_6_2 = f'''SELECT
                count(v.id)
                from vacancies v
                where (v.name ilike '%data scientist%'
                    or v.name ilike '%data science%'               
                    or v.name ilike '%исследователь данных%' 
                    or (v.name like '%ML%' and v.name not like '%HTML%')
                    or v.name ilike '%machine learning%'
                    or v.name ilike '%машинн%обучен%')
                and (v.name ilike '%junior%'
                    or v.experience = 'Нет опыта' or v.schedule = 'Стажировка')    
                '''

In [42]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\118162426.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
# считаем количесво вакансий с упомянутыми выше сочетаниями в названии и обязательно сочетаниями 'sql' либо 'postgres' в key_skills
query_6_3 = f'''SELECT count(vacancies.name) as n_vac
                FROM vacancies
                WHERE (lower(vacancies.name) LIKE '%data scientist%' 
                or lower(vacancies.name) LIKE '%data science%' 
                or lower(vacancies.name) LIKE '%исследователь данных%' 
                or lower(vacancies.name) LIKE '%ml%' and vacancies.name NOT LIKE '%HTML%' and vacancies.name not like '%QML%' 
                or lower(vacancies.name) LIKE '%machine learning%' 
                or lower(vacancies.name) LIKE '%машинн%обучен%')
                and(lower(vacancies.key_skills) LIKE '%postgres%'
                or lower(vacancies.key_skills) LIKE '%sql%')    
                '''

In [44]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\3310962512.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


n_vac
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
# считаем количесво вакансий с упомянутыми выше сочетаниями в названии и обязательно 'python' в key_skills
query_6_4 = f'''SELECT count(vacancies.name) as n_vac
                FROM vacancies
                WHERE (lower(vacancies.name) LIKE '%data scientist%' 
                or lower(vacancies.name) LIKE '%data science%' 
                or lower(vacancies.name) LIKE '%исследователь данных%' 
                or lower(vacancies.name) LIKE '%ml%' and vacancies.name NOT LIKE '%HTML%' and vacancies.name not like '%QML%' 
                or lower(vacancies.name) LIKE '%machine learning%' 
                or lower(vacancies.name) LIKE '%машинн%обучен%') 
                and lower(vacancies.key_skills) LIKE '%python%' 
                ''' 

In [46]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\614304167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


n_vac
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [47]:
# выбираем вакансии с упомянутыми выше сочетаниями в названии и обязательно заполненной графой key_skills, считаем среднее количество ключевых навыков
query_6_5 = f'''SELECT round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1), 2)
                FROM vacancies
                WHERE (vacancies.name ilike '%data scientist%' 
                or vacancies.name ilike '%data science%' 
                or vacancies.name ilike '%исследователь данных%' 
                or vacancies.name ilike '%ml%' and vacancies.name not ilike '%HTML%' 
                or vacancies.name ilike '%machine learning%' 
                or vacancies.name ilike '%машинн%обучен%')
                and vacancies.key_skills is not null  
                ''' 

In [48]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\4089566367.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


round
0   6.55

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [49]:
# выбираем вакансии с упомянутыми выше сочетаниями в названии, выделяем experience и считаем среднее значение Зарплаты по графам salary_from и salary_to
query_6_6 = f'''        
    SELECT experience,
    ROUND(avg(COALESCE((salary_to + salary_from)/2, salary_from, salary_to)), 0) AS result
    FROM vacancies 
    WHERE (name ILIKE '%data scientist%' OR name ILIKE '%data science%' OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%'
        )
    AND (salary_from is not null OR salary_to is not null)

    GROUP BY experience
    '''

In [50]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

C:\Users\Ирина\AppData\Local\Temp\ipykernel_31736\2297777544.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience    result
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

Вывод:
- Из 1771 вакансии, связаной с работой с данными, лишь 51 из них относится к сфере Data Science и подходит для junior специалиста.
- Из них 229 вакансий связаны с SQL и postgres.
- В вакансиях указано в среднем 6 пунктов в разделе ключевых навыков, при этом 359 вакансий указывают как ключевой навык язык Python.
- В среднем, для вакансий в сфере Data Science указана достаточно высокая ЗП, которая начинается от 74643 руб для специалистов без опыта работы. Самая высокая ЗП указана для специалистов с опытом 3 до 6 лет - 243115 руб. 

# Общий вывод по проекту

Рынок труда не равномерен - большинство вакансий из базы данных hh сконцентрированы в городах-миллионниках. Это логично, т.к. чем крупнее населёный пункт, тем больше в нём работодателей. 

Большинство работодателей ищут соискателей с опытом работы от 1 до 3 лет - то есть начинающих специалистов часто молодого возраста,  которые часто ещё меняют место работы, не обременены семьёй и не имеют сильно высоких ожиданий касательно размера заработной платы.

Многие работодатели не устанавливают в вакансиях размер ЗП, предположительно, чтобы получить больше откликов на вакансию. Так как человека легче уговорить на должность с ЗП меньше ожидаемой непосредственно во время интервью.

Основное количество размещенных вакансий принадлежит банкам и компаниям из отрасли информационных технологий - это работодатели, являющиеся основными потребителяи IT-специалистов. При этом почти треть работодателей не указывают сферу деятельности, нередко под личиной таких работодателей скрываются рекрутеры Яндекса и других крупных компаний, создавая иллюзию разнообразия рабтодателей.

Совсем небольшое количество вакансий Data Science подходят junior специалисту - работодатели в основном ищут сразу опытных работников, которые все сразу сами умеют. Обязательное владение языком Python тоже встречается сравнительно редко.

В сфере Data Science - высокая обещанная ЗП, которая даже для соискателей без опыта работы составляет от 70 000 и выше. 


